x/buyback/abci.go

```diff
@@ -17,24 +17,18 @@ func BeginBlocker(ctx sdk.Context, k Keeper, bk types.BankKeeper) {
 
 	var (
 		stakingDenom = k.GetStakingTokenDenom(ctx)
-		pricingInfo  = groupMarketDataBySource(k.GetMarketData(ctx), stakingDenom)
 		account      = k.GetBuybackAccountAddr()
 	)
 
 	for _, balance := range bk.GetAllBalances(ctx, account) {
-		pricedata, found := pricingInfo[balance.Denom]
-		if !found {
-			// do not have market data to create an order.
-			continue
-		}
-
-		if pricedata.LastPrice == nil {
-			// do not have market data to create an order.
+		price := k.GetBestPrice(ctx, balance.Denom, stakingDenom)
+		if price == nil {
+			// There are no passive orders to fill for this instrument
 			continue
 		}
 
 		// Calculate the amount of staking tokens that can be purchased at that price
-		destinationAmount := balance.Amount.ToDec().Mul(*pricedata.LastPrice).TruncateInt()
+		destinationAmount := balance.Amount.ToDec().Mul(*price).TruncateInt()
```

Comment:
As discussed, double check that TruncateInt is the same as floor(): DestinationAmount should be rounded down to make sure order is filled.

Reply:
It's confirmed.